osu.Game/Screens/Edit/Setup/ResourcesSection.cs

```diff
@@ -72,75 +74,92 @@ private void load()
         }
 
-        public bool ChangeBackgroundImage(FileInfo source)
+        public bool ChangeBackgroundImage(FileInfo source, bool applyToAllDifficulties)
         {
             if (!source.Exists)
                 return false;
 
-            var set = working.Value.BeatmapSetInfo;
-
-            var destination = new FileInfo($@"bg{source.Extension}");
-
-            // remove the previous background for now.
-            // in the future we probably want to check if this is being used elsewhere (other difficulties?)
-            var oldFile = set.GetFile(working.Value.Metadata.BackgroundFile);
-
-            using (var stream = source.OpenRead())
-            {
-                if (oldFile != null)
-                    beatmaps.DeleteFile(set, oldFile);
-
-                beatmaps.AddFile(set, stream, destination.Name);
-            }
+            changeResource(source, applyToAllDifficulties, @"bg",
+                metadata => metadata.BackgroundFile,
+                (metadata, name) => metadata.BackgroundFile = name);
 
-            editorBeatmap.SaveState();
-
-            working.Value.Metadata.BackgroundFile = destination.Name;
             headerBackground.UpdateBackground();
-
             editor?.ApplyToBackground(bg => bg.RefreshBackground());
-
             return true;
         }
 
-        public bool ChangeAudioTrack(FileInfo source)
+        public bool ChangeAudioTrack(FileInfo source, bool applyToAllDifficulties)
         {
             if (!source.Exists)
                 return false;
 
+            changeResource(source, applyToAllDifficulties, @"audio",
+                metadata => metadata.AudioFile,
+                (metadata, name) => metadata.AudioFile = name);
+
+            music.ReloadCurrentTrack();
+            return true;
+        }
+
+        private void changeResource(FileInfo source, bool applyToAllDifficulties, string baseFilename, Func<BeatmapMetadata, string> readFilename, Action<BeatmapMetadata, string> writeFilename)
+        {
             var set = working.Value.BeatmapSetInfo;
 
-            var destination = new FileInfo($@"audio{source.Extension}");
+            string newFilename = string.Empty;
 
-            // remove the previous audio track for now.
-            // in the future we probably want to check if this is being used elsewhere (other difficulties?)
-            var oldFile = set.GetFile(working.Value.Metadata.AudioFile);
+            if (applyToAllDifficulties)
+            {
+                newFilename = $"{baseFilename}{source.Extension}";
 
-            using (var stream = source.OpenRead())
+                foreach (var beatmap in set.Beatmaps)
+                {
+                    if (set.GetFile(readFilename(beatmap.Metadata)) is RealmNamedFileUsage otherExistingFile)
+                        beatmaps.DeleteFile(set, otherExistingFile);
+
+                    writeFilename(beatmap.Metadata, newFilename);
+                }
+            }
```

Comment:
This logic is missing a completely key step.

Recall that a beatmap's background filename is contained in the `.osu` file for the beatmap. None of this actually mutates the other `.osu` files. Which means that this breaks immediately on export and re-import; only the difficulty that was active at the point of replacing all backgrounds will have the backgrounds, and the other difficulties will point to dead assets that are no longer there because they were deleted at the point of replace. (I have tested this.)

In other words this logic needs to _re-encode every single beatmap in the set_ to work correctly.

Reply:
Should be addressed in https://github.com/ppy/osu/pull/30860/commits/8e0f6fc12dc04a224a9aefb0121f990a8b007af2, I don't like the part where I'm re-fetching each difficulty's skin in order to preserve it but can't see any other way around that.